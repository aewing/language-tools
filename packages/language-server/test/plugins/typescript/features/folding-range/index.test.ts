import { describe, it, expect, beforeAll } from 'vitest';
import { readdirSync, statSync, existsSync } from 'fs';
import { join } from 'path';
import ts from 'typescript';
import { Document, DocumentManager } from '../../../../../src/lib/documents';
import { LSConfigManager } from '../../../../../src/ls-config';
import { LSAndTSDocResolver } from '../../../../../src/plugins';
import { FoldingRangeProviderImpl } from '../../../../../src/plugins/typescript/features/FoldingRangeProvider';
import { pathToUrl } from '../../../../../src/utils';
import { serviceWarmup } from '../../test-utils';

function setup(workspaceDir: string, filePath: string) {
    const docManager = new DocumentManager(
        (textDocument) => new Document(textDocument.uri, textDocument.text)
    );
    const configManager = new LSConfigManager();
    configManager.updateClientCapabilities({
        textDocument: { foldingRange: { lineFoldingOnly: true } }
    });
    const lsAndTsDocResolver = new LSAndTSDocResolver(
        docManager,
        [pathToUrl(workspaceDir)],
        configManager
    );
    const plugin = new FoldingRangeProviderImpl(lsAndTsDocResolver, configManager);
    const document = docManager.openClientDocument(<any>{
        uri: pathToUrl(filePath),
        text: ts.sys.readFile(filePath) || ''
    });
    return { plugin, document, docManager, lsAndTsDocResolver };
}

describe('FoldingRangeProvider', () => {
    const fixturesDir = join(__dirname, 'fixtures');
    const workspaceDir = join(__dirname, '../../testfiles');

    beforeAll(() => {
        serviceWarmup(workspaceDir, pathToUrl(workspaceDir));
    });

    // Get all test fixtures
    const testFiles = readdirSync(fixturesDir).filter((entry) => {
        const fullPath = join(fixturesDir, entry);
        const inputFile = join(fullPath, 'input.svelte');
        return statSync(fullPath).isDirectory() && existsSync(inputFile);
    });

    for (const testName of testFiles) {
        it(testName, async () => {
            const inputFile = join(fixturesDir, testName, 'input.svelte');
            const { plugin, document } = setup(workspaceDir, inputFile);
            const folding = await plugin.getFoldingRanges(document);

            expect(folding).toMatchSnapshot();
        });
    }
});
