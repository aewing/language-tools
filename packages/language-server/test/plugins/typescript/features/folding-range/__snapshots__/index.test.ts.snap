// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FoldingRangeProvider > await-catch 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 7,
    "startLine": 6,
  },
]
`;

exports[`FoldingRangeProvider > await-peding-catch 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 3,
    "startLine": 2,
  },
  {
    "endLine": 7,
    "startLine": 6,
  },
  {
    "endLine": 10,
    "startLine": 9,
  },
]
`;

exports[`FoldingRangeProvider > await-pending-only 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
]
`;

exports[`FoldingRangeProvider > await-pending-then 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 3,
    "startLine": 2,
  },
]
`;

exports[`FoldingRangeProvider > await-pending-then-catch 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 3,
    "startLine": 2,
  },
  {
    "endLine": 5,
    "startLine": 4,
  },
  {
    "endLine": 9,
    "startLine": 8,
  },
  {
    "endLine": 12,
    "startLine": 11,
  },
  {
    "endLine": 15,
    "startLine": 14,
  },
]
`;

exports[`FoldingRangeProvider > await-then 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 7,
    "startLine": 6,
  },
]
`;

exports[`FoldingRangeProvider > await-then-catch 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 3,
    "startLine": 2,
  },
]
`;

exports[`FoldingRangeProvider > await-then-catch-shorthand 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 3,
    "startLine": 2,
  },
]
`;

exports[`FoldingRangeProvider > each-block 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 6,
    "startLine": 5,
  },
]
`;

exports[`FoldingRangeProvider > each-else 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 3,
    "startLine": 2,
  },
]
`;

exports[`FoldingRangeProvider > each-keyed 1`] = `
[
  {
    "endLine": 3,
    "startLine": 2,
  },
]
`;

exports[`FoldingRangeProvider > if-block 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 3,
    "startLine": 2,
  },
  {
    "endLine": 5,
    "startLine": 4,
  },
  {
    "endLine": 9,
    "startLine": 8,
  },
  {
    "endLine": 13,
    "startLine": 12,
  },
  {
    "endLine": 15,
    "startLine": 14,
  },
]
`;

exports[`FoldingRangeProvider > key-block 1`] = `
[
  {
    "endLine": 1,
    "startLine": 0,
  },
  {
    "endLine": 7,
    "startLine": 6,
  },
]
`;

exports[`FoldingRangeProvider > parser-error 1`] = `
[
  {
    "endCharacter": undefined,
    "endLine": 2,
    "kind": undefined,
    "startCharacter": undefined,
    "startLine": 1,
  },
  {
    "endLine": 7,
    "startLine": 6,
  },
  {
    "endLine": 9,
    "startLine": 8,
  },
]
`;

exports[`FoldingRangeProvider > script 1`] = `
[
  {
    "endCharacter": undefined,
    "endLine": 2,
    "kind": "imports",
    "startCharacter": undefined,
    "startLine": 1,
  },
  {
    "endCharacter": undefined,
    "endLine": 5,
    "kind": undefined,
    "startCharacter": undefined,
    "startLine": 4,
  },
]
`;
